compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/TailrecLowering.kt

```diff
@@ -156,25 +157,22 @@ private class BodyTransformer(
             // that captures another parameter, it won't capture it as a mutable ref.
             parameter to irTemporary(argument)
         }
-
-        // For each unspecified argument set the corresponding variable to default:
-        parameters
-            .filter { it !in parameterToArgument }
-            .let { if (properComputationOrderOfTailrecDefaultParameters) it else it.asReversed() }
+        // Create new null-initialized variables for all other values in case of forward references:
+        //   fun f(x: () -> T = { y }, y: T = ...) // in `f()`, `x()` returns `null`
+        val defaultValuedParameters = parameters.filter { it !in parameterToArgument }
+        defaultValuedParameters.associateWithTo(parameterToArgument) {
+            // Note that we intentionally keep the original type of the parameter for the variable even though that violates type safety
+            // if it's non-null. This ensures that capture parameters have the same types for all copies of `x`.
+            // TODO: on the JVM, this might not be correct for inline classes - copy code from JvmDefaultParameterInjector?
```

Comment:
@ilmirus @dnpetrov Could you come up with a test ?

Reply:
That seems to work fine. The class should have a primitive underlying type:
```
@JvmInline
value class C(val x: Int)

tailrec fun test(flag: Boolean, fn: () -> C = { y }, y: C = C(1)): Int =
    if (flag) fn().x else test(true)

fun box() = test(false)
```
That way the unsafe coerce intrinsic actually does something.

Reply:
Correction: the problem is not the unsafe coerce intrinsic, but that IrConstImpl.defaultValueForType will produce null when given type C instead of 0. In fact, calling `makeNullable` here was wrong.